lib/statsd/instrument/aggregator.rb

```diff
@@ -0,0 +1,246 @@
+        unless thread_healthcheck
+          sink << datagram_builder(no_prefix: no_prefix).g(name, value, CONST_SAMPLE_RATE, tags)
+          return
+        end
+
+        tags = tags_sorted(tags)
+        key = packet_key(name, tags, no_prefix, COUNT)
+
+        mutex.synchronize do
+          unless aggregation_state.key?(key)
+            aggregation_state[key] = {
+              type: GAUGE,
+              name: name,
+              value: value,
+              tags: tags,
+              no_prefix: no_prefix,
+            }
+          end
+          # keep the last value
+          aggregation_state[key][:value] = value
+        end
+      end
+
+      def flush
+        mutex.synchronize(&method(:do_flush))
+      end
+
+      private
+
+      def do_flush
+        aggregation_state.each do |_key, agg|
+          case agg[:type]
+          when COUNT
+            sink << datagram_builder(no_prefix: agg[:no_prefix]).c(
+              agg[:name],
+              agg[:value],
+              CONST_SAMPLE_RATE,
+              agg[:tags],
+            )
+          when DISTRIBUTION, MEASURE, HISTOGRAM
+            sink << datagram_builder(no_prefix: agg[:no_prefix]).distribution_value_packed(
+              agg[:name],
+              agg[:type].to_s,
+              agg[:value],
+              CONST_SAMPLE_RATE,
+              agg[:tags],
+            )
+          when GAUGE
+            sink << datagram_builder(no_prefix: agg[:no_prefix]).g(
+              agg[:name],
+              agg[:value],
+              CONST_SAMPLE_RATE,
+              agg[:tags],
+            )
+          else
+            StatsD.logger.error { "[#{self.class.name}] Unknown aggregation type: #{agg[:type]}" }
+          end
+        end
+        aggregation_state.clear
+      end
+
+      attr_reader :mutex, :aggregation_state, :sink
+
+      def tags_sorted(tags)
+        return [].freeze if tags.nil? || tags.empty?
+
+        if tags.is_a?(Hash)
+          sorted = tags.sort_by { |k, _v| k.to_s }
+          tags = sorted.map! { |k, v| "#{k}:#{v}" }
+        else
+          tags.sort!
+        end
+        tags
+      end
+
+      def packet_key(name, tags = [], no_prefix = false, type = COUNT)
+        "#{name}#{tags.join}#{no_prefix}#{type}".b
+      end
+
+      def datagram_builder(no_prefix:)
+        @datagram_builders[no_prefix] ||= @datagram_builder_class.new(
+          prefix: no_prefix ? nil : @metric_prefix,
+          default_tags: @default_tags,
+        )
+      end
+
+      def thread_healthcheck
+        unless @flush_thread&.alive?
+          return false unless Thread.main.alive?
+
+          if @pid != Process.pid
+            StatsD.logger.info { "[#{self.class.name}] Restarting the flush thread after fork" }
+            @pid = Process.pid
+            @aggregation_state.clear
+          else
+            StatsD.logger.info { "[#{self.class.name}] Restarting the flush thread" }
+          end
+          @flush_thread = Thread.new do
```

Comment:
This is called from unsynchonized places. All the `unless thread_healthcheck` may cause a race condition.

Reply:
I put this whole method under synchronization, but now I am thinking it might be very expensive. Let's see.